src/flb_pack_gelf.c

```diff
@@ -548,17 +549,38 @@ flb_sds_t flb_msgpack_to_gelf(flb_sds_t *s, msgpack_object *o,
                 if (v->type == MSGPACK_OBJECT_POSITIVE_INTEGER) {
                     if ( v->via.u64 > 7 ) {
                         flb_error("[flb_msgpack_to_gelf] level is %" PRIu64 ", "
-                                  "but should be in 0..7", v->via.u64);
+                                  "but should be in 0..7 or a syslog keyword", v->via.u64);
                         return NULL;
                     }
                 }
                 else if (v->type == MSGPACK_OBJECT_STR) {
                     val     = v->via.str.ptr;
                     val_len = v->via.str.size;
-                    if (val_len != 1 || val[0] < '0' || val[0] > '7') {
-                        flb_error("[flb_msgpack_to_gelf] level is '%.*s', "
-                                  "but should be in 0..7", val_len, val);
-                        return NULL;
+                    if (val_len == 1 && val[0] >= '0' && val[0] <= '7') {
+                        v = &vtmp;
+                        v->type = MSGPACK_OBJECT_POSITIVE_INTEGER;
+                        v->via.u64 = (uint64_t)(val[0] - '0');
+                    }
+                    else {
+                        int i;
+                        char* allowed_levels[] = {
+                            "emerg", "alert", "crit", "err",
+                            "warning", "notice", "info", "debug",
+                            NULL
+                        };
+                        for (i = 0; allowed_levels[i] != NULL; ++i) {
+                            if (!strncasecmp(val, allowed_levels[i], strlen(allowed_levels[i]))) {
```

Comment:
Since msgpack string is not null terminated the third argument of `strncasecmp` should be val_len, if not you can be reading bytes out of the msgpack buffer.

Also you are right, we need to check the length of the strings:
```
if ((val_len == strlen(allowed_levels[i])) && !strncasecmp(val, allowed_levels[i], val_len))
```

Reply:
Okay, I'll add this asap.
But just for curiosity, the reason that you mentioned they're not null terminated is about possible page faults, or about security reasons?

Reply:
Yes, you can get a page fault. 
The buffer over-read it's a security problem, the implications depends on the case.
 https://cwe.mitre.org/data/definitions/126.html